### ISSUE TITLE
Type and Member filtering for running specs

### ISSUE THREAD

**Issue by einari** [machine/machine.specifications#313]
Add support for -filters for filtering down to specific types to run - similar to what the console runner for the .NET Framework has. 

``` terminal
dotnet test --filters My.Spec.In.Namespace.MySpec, My.Spec.In.Namespace.MyOtherSpec
```

In addition, it would be really nice to have something to specify which member within a specification should be run. 

``` terminal
dotnet test --specs My.Spec.In.Namespace.MySpec#Member, My.Spec.In.Namespace.MyOtherSpec#Member
```

This will be extremely helpful in the Visual Studio Code extension I'm working on. I plan to invoke the `dotnet test` tool directly.

**Comment by BrunoJuchli** [comment]
Could you elaborate what you mean by "member within a specification"?

**Comment by einari** [comment]
The actual spec (The IT statement). 

The VSCode extension I'm building will be context sensitive in the sense that you can be on a specification assertion and run that spec only.

**Comment by BrunoJuchli** [comment]
Should it also support evaluating a subset (meaning m of n) of `It`s of the same spec? In that case it'd be preferrable to run the `Establish` and `Because` only once (as of now).

**Comment by einari** [comment]
Sure. I'm thinking of invoking the RunMembers() of the test Controller. I believe the internal runner then will do the right thing.

**Comment by ivanz** [comment]
@einari Have a look at the test [Controller](https://github.com/machine/machine.specifications/blob/master/Source/Machine.Specifications/Controller/Controller.cs) and how it is used in dotnet test [here](https://github.com/machine/dotnet-test-mspec/blob/master/src/dotnet-test-mspec/Controller/TestController.cs). I've added support for running a subset of `It` in 0.11 . The interface is documented here: https://github.com/machine/machine.specifications/wiki/Test-Controller-Protocol

Tests for it [are here](https://github.com/machine/machine.specifications/blob/master/Source/Machine.Specifications.Specs/Controller/ControllerSpecs.cs)

**Comment by ivanz** [comment]
@einari  Another thing to note is that I am not sure if there is already a "--tests" param used by other runners - would be good to have a consistent param across all the runners 👍

**Comment by einari** [comment]
Excellent - thanks for the info. 
I can implement the same feature in the console runner for .NET Framework to make it consistent across the board. Easier than having to implement a specific console app for VSCode to be able to do this.. (I might be considered lazy at this point.. :) )

**Comment by einari** [comment]
Btw: What is preferred "--tests" or "--specs"  ?

**Comment by ivanz** [comment]
When I meant the other runners - I was referring to xunit and nunit' dotnet test runners - not sure which flag they use.

**Comment by einari** [comment]
From the looks of it NUnit has a couple of options, but the one we're looking for is probably the `list`option: 

```csharp
this.Add("list", "Used by IDEs to request a list of tests that can be run",  v => List = v != null);
```

from (https://github.com/nunit/dotnet-test-nunit/blob/master/src/dotnet-test-nunit/CommandLineOptions.cs)[NUnits repo]

**Comment by einari** [comment]
This vacation thing has really done my head in - obviously it should have been the ´-test / --test´ option:

```csharp
this.Add("test=", "Comma-separated list of {NAMES} of tests to run or explore. This option may be repeated.",
                v => ((List<string>)TestList).AddRange(TestNameParser.Parse(RequiredValue(v, "--test"))));
```

I'll dig into the format of what is being sent to these so that we match that.

**Comment by einari** [comment]
I've now implemented something. But please let me know if this is just wrong. I've been zipping through the Nunit and Xunit runners and they both implement -test and basically from what I can gather they both rely on unique testnames. The way I write my MSpec Specifications, there is no way to guarantee a unique name without including the entire namespace for the spec, because I don't prefix or postfix the classname - or include in any way a reference to the unit being specified. So I figured that -test takes a comma separated list of fully qualified Namespace + class and an optional # and then the name of a specific assertion. 

Anyways, PR sitting here: https://github.com/machine/dotnet-test-mspec/pull/6
